app/src/main/java/pl/droidevs/books/exportimport/ExportImportViewModel.java

```diff
@@ -48,28 +47,26 @@ public void onChanged(@Nullable List<BookEntity> books) {
                 try {
                     writeToFile(file, books);
                 } catch (IOException e) {
-                    wasSavingSuccessful[0] = false;
+                    throw new ExportFailedException(e.getCause());
```

Comment:
Używając `e.getCause()` tracisz prawdziwe źródło wyjątku. Spróbuj uruchomić poniższy test:
`@Test
    public void test() {
        try {
            Scanner in = new Scanner(new FileReader("filename.txt"));
        } catch (IOException e) {
            try {
                throw new ExportFailedException(e.getCause());
            } catch (ExportFailedException ef) {
                ef.printStackTrace();
            }
        }
    }`
Gdy użyjesz `ExportFailedException(e.getCause())` Twój stacktrace będzie wyglądał tak:

> pl.droidevs.books.ExampleUnitTest$ExportFailedException
	at pl.droidevs.books.ExampleUnitTest.test(ExampleUnitTest.java:29)

Jednak gdy przekażesz obiekt wyjątku `ExportFailedException(e.)` to później łatwiej jest określić co poszło nie tak:

> pl.droidevs.books.ExampleUnitTest$ExportFailedException: java.io.FileNotFoundException: filename.txt (No such file or directory)
	at pl.droidevs.books.ExampleUnitTest.test(ExampleUnitTest.java:29)